src/dioptra/restapi/v1/jobs/service.py

```diff
@@ -527,6 +525,172 @@ def get(
+        Returns:
+            The metric dictionary passed in if successful, otherwise an error message.
+        """
+        log: BoundLogger = kwargs.get("log", LOGGER.new())
+        log.debug("Update job metrics by id", job_id=job_id)
+
+        from mlflow.tracking import MlflowClient
+
+        run_id: UUID | None = self._job_id_mlflowrun_service.get(
+            job_id=job_id, **kwargs
+        )["mlflow_run_id"]
+
+        if run_id is None:
+            raise EntityDoesNotExistError("MlFlowRun", run_id=None)
+        else:
+            client = MlflowClient()
+
+            # this is here just to raise an error if the run does not exist
+            try:
+                run = client.get_run(run_id.hex)  # noqa: F841
+            except MlflowException as e:
+                raise EntityDoesNotExistError("MlFlowRun", run_id=run_id.hex) from e
+
+            client.log_metric(
+                run_id.hex,
+                key=metric_name,
+                value=metric_value,
+                step=metric_step,
+            )
+        return {"name": metric_name, "value": metric_value}
+
+
+class JobIdMetricsSnapshotsService(object):
+    """The service methods for retrieving the historical metrics of a
+    job by unique id and metric name."""
+
+    @inject
+    def __init__(self, job_id_mlflowrun_service: JobIdMlflowrunService) -> None:
+        """Initialize the job metrics snapshots service.
+
+        All arguments are provided via dependency injection.
+        """
+        self._job_id_mlflowrun_service = job_id_mlflowrun_service
+
+    def get(
+        self,
+        job_id: int,
+        metric_name: str,
+        page_index: int,
+        page_length: int,
+        **kwargs,
+    ) -> tuple[list[dict[str, Any]], int]:
+        """Fetch a job's metrics by its unique id and metric name.
+
+        Args:
+            job_id: The unique id of the job.
+            metric_name: The name of the metric.
+            page_index: The index of the first page to be returned.
+            page_length: The maximum number of experiments to be returned.
+        Returns:
+            The metric history for the requested job and metric if found,
+            otherwise an error message.
+        """
+        from mlflow.tracking import MlflowClient
+
+        log: BoundLogger = kwargs.get("log", LOGGER.new())
+        log.debug(
+            "Get job metric history by id and name",
+            job_id=job_id,
+            metric_name=metric_name,
+        )
+
+        run_id: UUID | None = self._job_id_mlflowrun_service.get(
+            job_id=job_id, **kwargs
+        )["mlflow_run_id"]
+        client = MlflowClient()
+
+        if run_id is None:
+            raise EntityDoesNotExistError("MlFlowRun", run_id=None)
+
+        try:
+            history = client.get_metric_history(run_id=run_id.hex, key=metric_name)
+        except MlflowException as e:
+            raise EntityDoesNotExistError("MlFlowRun", run_id=run_id.hex) from e
```

Comment:
How do we want to handle errors here?

`get_metric_history` returns an empty list if the `run_id` is not found or the there are no metrics with `key`. It doesn't raise an error. So we don't actually get an error raised by mlflow. It does raise an error if `run_id` is None, but it's a `INVALID_PARAMETER_VALUE_ERROR`

Do we want to raise an error if the metric with the specified name is not found? When we retrieve all metrics and there are none, we simply return an empty list, but maybe requesting a specific name should operate differently. Should this be a new error type? I thought `EntityDoesNotExist` was more for top level entities (e.g. Resources, Users, Groups).

We still might want a try/except around any mlflow client invocations to catch any kind of connectivity errors. Should we have some new error type for errors with an internal dioptra service?

Reply:
@jkglasbrenner @jtsextonMITRE

Reply:
@jkglasbrenner these questions about error handling are my remaining unresolved issues with this PR. Once we resolve this I would approve the PR.